onchange_helper/models/models.py

```diff
@@ -43,9 +43,24 @@ def play_onchanges(self, values, onchange_fields):
         for field in onchange_fields:
             onchange_values = self.onchange(all_values, field, onchange_specs)
             new_values.update(self._get_new_values(values, onchange_values))
-            all_values.update(new_values)
+            clean_new_values = self.clean_o2m_m2o_m2m_fields(new_values)
+            all_values.update(clean_new_values)
 
         return {
             f: v for f, v in all_values.iteritems()
             if not self._fields[f].compute
             and (f in values or f in new_values)}
+
+    @api.multi
+    def clean_o2m_m2o_m2m_fields(self, vals):
+        for itens in vals:
```

Comment:
this code is hard to read, can you add a docstring explaining what you want to achieve ?

Reply:
Hi @hparfr , thanks for review. Basically I try to solve the problem describe in issue https://github.com/OCA/server-tools/issues/1457 . The dict return by the method has wrong values for m2o fields, what cause error, the dict must be treat

the dict returned is

{'tax_line_ids': [(5,), (0, 0, {'account_id': (14, u'111200 Tax Received'), 'sequence': 1, 'manual': False, 'name': u'Tax 15.00%', 'currency_id': (3, u'USD'), 'amount': 296.40000000000003, 'account_analytic_id': False, 'tax_id': (1, u'Tax 15.00%')})], 'invoice_line_ids': [(5,), (0, 0, {'uom_id': (1, u'Unit(s)'), 'account_id': (5, u'101120 Stock Interim Account (Received)'), 'analytic_tag_ids': [(5,)], 'move_line_id': (38, u'Test - account_invoice_from_picking/E-COM05: Stock>Customers'), 'price_unit': 247.0, 'sequence': 0, 'price_subtotal': 1976.0, 'company_id': (1, u'YourCompany'), 'currency_id': (3, u'USD'), 'discount': 0.0, 'product_id': (14, u'[E-COM05] Bose Mini Bluetooth Speaker'), 'account_analytic_id': False, 'invoice_line_tax_ids': [(5,), (4, 1)], 'quantity': 8.0, 'picking_id': (28, u'WH/OUT/00014'), 'name': u'Bose Mini Bluetooth Speaker'})], 'picking_id': 28}

The fields o2m need to be id not a list ( 'currency_id': (3, u'USD') ), if you try to write the values in this way you get error message, the dict should be like this

{'tax_line_ids': [(5,), (0, 0, {'account_id': 14, 'sequence': 1, 'manual': False, 'name': u'Tax 15.00%', 'currency_id': 3, 'amount': 296.40000000000003, 'account_analytic_id': False, 'tax_id': 1})], 'invoice_line_ids': [(5,), (0, 0, {'uom_id': 1, 'account_id': 5, 'analytic_tag_ids': [(5,)], 'move_line_id': 38, 'price_unit': 247.0, 'sequence': 0, 'price_subtotal': 1976.0, 'company_id': 1, 'currency_id': 3, 'discount': 0.0, 'product_id': 14, 'account_analytic_id': False, 'invoice_line_tax_ids': [(5,), (4, 1)], 'quantity': 8.0, 'picking_id': 28, 'name': u'Bose Mini Bluetooth Speaker'})], 'picking_id': 28}